### PR TITLE
port write buffer manager and compaction thread limiter

### PR DESCRIPTION
**Pull request by tabokie**
For https://github.com/tikv/rocksdb/pull/305 and https://github.com/facebook/rocksdb/pull/4332.

Also, refactor rate limiter for sharing between instances.

### CODE REVIEW

File: src/rocksdb_options.rs [hunk 1]
```diff
@@ -414,6 +421,61 @@ impl Drop for Statistics {
     }
 }
 
+pub struct WriteBufferManager {
+    pub inner: *mut DBWriteBufferManager,
```
Comment (BusyJay) [inline src/rocksdb_options.rs:425]: Why pub?

**Review by BusyJay — APPROVED** [review]
Rest LGTM

File: src/rocksdb_options.rs (hunk 1 above)
Comment (tabokie) [reply src/rocksdb_options.rs:425]: It's used in `crocksdb_options_set_write_buffer_manager`, I'll change them to pub(crate) instead.

**Review by tonyxuqqi — APPROVED** [review]

**Review by tonyxuqqi — APPROVED** [review]

File: .gitmodules [hunk 2]
```diff
@@ -1,7 +1,7 @@
 [submodule "rocksdb"]
 	path = librocksdb_sys/rocksdb
-	url = https://github.com/tikv/rocksdb.git
-	branch = 6.29.tikv
+	url = https://github.com/tabokie/rocksdb.git
+	branch = schedule-memtable
```
Comment (tonyxuqqi) [inline .gitmodules:4]: When to merge the rocksdb code to tikv branch.

**Comment by tonyxuqqi** [comment]
get/set_rate_bytes_per_sec, get/set_auto_tuned, get_and_reset_statistics_ticker_count, 
get_statistics_histogram
reset_statistics

These methods are needed to compile the current TiKV node.

**Comment by tabokie** [comment]
@tonyxuqqi The changes are by design. See https://github.com/tikv/tikv/pull/13984

**Pull request merged**